### ISSUE TITLE
Problem with rejoin  'join_request_fail', problem with maxClients value

### ISSUE THREAD

**Issue by daimonkor** [colyseus/colyseus#224]
Hi all, I found some strange state when use rejoin feature

My server  "colyseus": "^0.9.32",

```js
import {Presence, Room} from "colyseus";

enum State {
    WAITING, GAME, LOST_PLAYER, NONE
}

export class GameState {
    masterPlayer = null;
    slavePlayer = null;
    state = State.NONE;
    command = {};

    withState(state: State){
        this.state = state;
        return this;
    }
}

class RoomData {
    name: String;
    password: String;

    constructor(name: String, password: String) {
        this.name = name;
        this.password = password;
    }
}

class AuthData {
    token: String;
    password: String;

    constructor(token: String, password: String) {
        this.token = token;
        this.password = password;
    }
}

export class BattleRoom extends Room<GameState> {
    maxClients = 2;
    autoDispose = true;

    constructor(presence?: Presence) {
        super(presence);
    }

    requestJoin(options, isNew) {
        console.log(this.getLogTag(), "requestJoin", options, isNew);
        return (options.create)
            ? (options.create && isNew)
            : this.clients.length > 0;
    }

    onInit(options) {
        console.log(this.getLogTag(), "Game Room created!", options);
        this.setSeatReservationTime(20);
        this.setState(new GameState().withState(State.WAITING));
        this.metadata = new RoomData(options.name, options.password)
    }

    onAuth(options): boolean | Promise<any>  {
        console.log(this.getLogTag(), options.password, this.metadata);
        return this.checkAuth(new AuthData(null, options.password))
    }

    onJoin(client, options?, auth?) {
        console.log(this.getLogTag(), options, auth, client.id);
        if (options.create) {
            this.state.masterPlayer = client.id
        } else {
            this.state.slavePlayer = client.id
        }
        console.log(this.getLogTag(), 'Client joined: ' + client.id);
        if (this.clients.length == 2) {
            this.state.state = State.GAME;
            let command = {};
            this.clients.forEach(function (value) {
                command[value.id] = {}
            });
            this.state.command = command
        }
        this.broadcastPatch()
    }

    async onLeave(client, consented?) {
        try {
            if (consented) {
                throw new Error("consented leave");
            }

            console.log(this.getLogTag(), "Count users:", this.clients.length, consented);
            await this.allowReconnection(client);
            console.log(this.getLogTag(), 'Reconnect client: ' + client.id);
        } catch (e) {
            console.log(this.getLogTag(), 'Error: ' + e);
            console.log(this.getLogTag(), 'Client left: ' + client.id);

            if (client.id == this.state.masterPlayer) {
                this.state.masterPlayer = null;
             //   this.unlock()

            } else if (client.id == this.state.slavePlayer) {
                this.state.slavePlayer = null;
         //       this.unlock()
            }
        }
    }

    getLogTag() {
        return '[' + (new Date()).toLocaleString() + '] ' + this.roomName + '(' + this.roomId + '):';
    }

    private checkAuth(auth?: AuthData): boolean {
        console.log(this.getLogTag(), this.metadata.password, auth.password)
        return auth.password == this.metadata.password
    }
}
```

Client part:
```kotlin

   private val serverRoomListener = object : Room.Listener() {

        override fun onLeave() {
            try{
                Thread.sleep(5000)
                ServerClient.getInstance(ServerClient.Builder(url = SERVER_WS)).also {
                    reconnect = true
                    it.connect()
                }
            }catch(e: java.lang.Exception){
                Timber.e(e)
            }
        }
}

    private val serverClientListener = object : Client.Listener {

        override fun onOpen(id: String?) {
            Timber.e("$id ${currentRoomWrapper.name} ${currentRoomWrapper.roomId}")
            if(reconnect){
                ServerClient.getInstance(ServerClient.Builder(url = SERVER_WS))
                        .join(currentRoomWrapper.name, linkedMapOf("sessionId" to currentRoomWrapper.roomId, "create" to false, "password" to currentRoomWrapper.password))
                     //   .rejoin(currentRoomWrapper.name, currentRoomWrapper.roomId)
                        .addListener(serverRoomListener)
                reconnect= false
            }
        }
}
```
found server exception https://prnt.sc/mw1m1g

**Important** 
* when I changed **maxClient = 3** http://prntscr.com/mw1qoe
* I used rejoin function (`roomName` and `sessionId`) and got http://prntscr.com/mw1ru1. Problem reproduced when `options` object is null. I don't know it is valid or no, but I tryied reconnect to room, no connect as new user, but other user can use `roomName` and `sessionId` too. 

Research:
* https://prnt.sc/mw2fc7 `_reserveSeat` calling without `allowReconnection = true`
* when `allowReconnection = false` run check on count clients https://prnt.sc/mw2eum
*  https://prnt.sc/mw2f0o `hasReachedMaxClients` is sum alive and reserved clients, mys state equal 2 (my room class `maxClient = 2`), because I have first client as alive and second as reserved 

Thanks for support.

**Comment by daimonkor** [comment]
Hi, @endel, please check rejoin Colyseus feature https://prnt.sc/mw1m1g. Thanks for your support.

**Comment by endel** [comment]
Hi @daimonkor, are you still using Colyseus? I've recently released version `0.11`, and this issue you've been facing shouldn't happen anymore as the seat reservation has been improved. 

- The `.rejoin()` method has been renamed to `.reconnect()`: https://docs.colyseus.io/client/client/#reconnect-roomid-string-sessionid-string
- Please see full migration guide: https://docs.colyseus.io/migrating/0.11/
